haystack_experimental/components/tools/tool_invoker.py

```diff
@@ -0,0 +1,242 @@
+
+    def _prepare_tool_result_message(self, result: Any, tool_call: ToolCall) -> ChatMessage:
+        """
+        Prepares a ChatMessage with the result of a tool invocation.
+
+        :param result:
+            The tool result.
+        :returns:
+            A ChatMessage object containing the tool result as a string.
+
+        :raises
+            StringConversionError: If the conversion of the tool result to a string fails
+            and `raise_on_failure` is True.
+        """
+        error = False
+
+        if self.convert_result_to_json_string:
+            try:
+                # We disable ensure_ascii so special chars like emojis are not converted
+                tool_result_str = json.dumps(result, ensure_ascii=False)
+            except Exception as e:
+                if self.raise_on_failure:
+                    raise StringConversionError("Failed to convert tool result to string using `json.dumps`") from e
+                tool_result_str = _TOOL_RESULT_CONVERSION_FAILURE.format(error=e, conversion_function="json.dumps")
+                error = True
+            return ChatMessage.from_tool(tool_result=tool_result_str, error=error, origin=tool_call)
+
+        try:
+            tool_result_str = str(result)
+        except Exception as e:
+            if self.raise_on_failure:
+                raise StringConversionError("Failed to convert tool result to string using `str`") from e
+            tool_result_str = _TOOL_RESULT_CONVERSION_FAILURE.format(error=e, conversion_function="str")
+            error = True
+        return ChatMessage.from_tool(tool_result=tool_result_str, error=error, origin=tool_call)
+
+    @component.output_types(tool_messages=List[ChatMessage])
+    def run(self, messages: List[ChatMessage]) -> Dict[str, Any]:
+        """
+        Processes ChatMessage objects containing tool calls and invokes the corresponding tools, if available.
+
+        :param messages:
+            A list of ChatMessage objects.
+        :returns:
+            A dictionary with the key `tool_messages` containing a list of ChatMessage objects with tool role.
+            Each ChatMessage objects wraps the result of a tool invocation.
+
+        :raises ToolNotFoundException:
+            If the tool is not found in the list of available tools and `raise_on_failure` is True.
+        :raises ToolInvocationError:
+            If the tool invocation fails and `raise_on_failure` is True.
+        :raises StringConversionError:
+            If the conversion of the tool result to a string fails and `raise_on_failure` is True.
+        """
+        tool_messages = []
+
+        for message in messages:
+            tool_calls = message.tool_calls
+            if not tool_calls:
+                continue
+
+            for tool_call in tool_calls:
+                tool_name = tool_call.tool_name
+                tool_arguments = tool_call.arguments
+
+                if not tool_name in self._tools_with_names:
```

Comment:
small nitpick for pep 8 standard
```suggestion
                if tool_name not in self._tools_with_names:
```

Reply:
Yeah sorry I just realized that a few min ago :D 

Just so I understand why can't we use the ToolInvoker from Haystack directly if this is a straight copy? Is there any need to copy it in haystack-experimental at all?

Reply:
As discussed offline, there are changes coming to this code with #175 but to keep the diff small we copy over the relevant components first so that only the changes need to be reviewed.